tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -887,7 +889,104 @@ class SPIRVEmitter : public ASTConsumer {
     theBuilder.setInsertPoint(mergeBB);
   }
 
-  void doForStmt(const ForStmt *forStmt) {
+  // TODO: add attributes (roll and unroll)
+  void doWhileStmt(const WhileStmt *whileStmt,
+                   llvm::ArrayRef<const Attr *> attrs = {}) {
+    // While loops are composed of:
+    //   while (<check>)  { <body> }
+    //
+    // SPIR-V requires loops to have a merge basic block as well as a continue
+    // basic block. Even though while loops do not have an explicit continue
+    // block as in for-loops, we still do need to create a continue block.
+    //
+    // Since SPIR-V requires structured control flow, we need two more basic
+    // blocks, <header> and <merge>. <header> is the block before control flow
+    // diverges, and <merge> is the block where control flow subsequently
+    // converges. The <check> block can take the responsibility of the <header>
+    // block. The final CFG should normally be like the following. Exceptions
+    // will occur with non-local exits like loop breaks or early returns.
+    //
+    //            +----------+
+    //            |  header  | <------------------+
+    //            | (check)  |                    |
+    //            +----------+                    |
+    //                 |                          |
+    //         +-------+-------+                  |
+    //         | false         | true             |
+    //         |               v                  |
+    //         |            +------+     +------------------+
+    //         |            | body | --> | continue (no-op) |
+    //         v            +------+     +------------------+
+    //     +-------+
+    //     | merge |
+    //     +-------+
+    //
+    // For more details, see "2.11. Structured Control Flow" in the SPIR-V spec.
+
+    spv::LoopControlMask loopControl = spv::LoopControlMask::MaskNone;
+    if (!attrs.empty()) {
+      if (attrs.front()->getKind() == attr::HLSLLoop)
```

Comment:
I typically prefer `switch` for checking all variants of a POD variable. This way you don't need to evaluate the variable multiple times. (Although the compiler may optimize them into one evaluation.) And it will also make things clear that we do nothing for `fastopt`. Give unimplemented error for `allow_uav_condition` maybe?